<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="MOFAcellulaR">
<title>Running a multicellular factor analysis in a cross-condition single-cell atlas • MOFAcellulaR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Running a multicellular factor analysis in a cross-condition single-cell atlas">
<meta property="og:description" content="MOFAcellulaR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">MOFAcellulaR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/get-started.html">Running a multicellular factor analysis in a cross-condition single-cell atlas</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Running a multicellular factor analysis in a cross-condition single-cell atlas</h1>
            
      
      
      <div class="d-none name"><code>get-started.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://saezlab.github.io/MOFAcellulaR/">MOFAcellulaR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="multicellular-factor-analysis">Multicellular factor analysis<a class="anchor" aria-label="anchor" href="#multicellular-factor-analysis"></a>
</h2>
<p>We repurposed the statistical framework of multi-omics factor
analysis <a href="https://www.embopress.org/doi/full/10.15252/msb.20178124" class="external-link">(MOFA)</a>
and <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02015-1" class="external-link">MOFA+</a>
to analyze cross-condition single cell atlases. These atlases profile
molecular readouts (eg. gene expression) of individual cells per sample
that can be classified into groups based on lineage (cell types) or
functions (cell states). We assumed that this nested design could be
represented as a multi-view dataset of a collection of patients, where
each individual view contains the summarized information of all the
features of a cell type per patient (eg. pseudobulk). In this data
representation there can be as many views as cell types in the original
atlas. MOFA then is used to estimate a latent space that captures the
variability of patients across the distinct cell types. The estimated
factors composing the latent space can be interpreted as a multicellular
program that captures coordinated expression patterns of distinct cell
types. The cell type specific gene expression patterns can be retrieved
from the factor loadings, where each gene of each cell type would
contain a weight that contributes to the factor score. Similarly, as in
the application of MOFA to multiomics data, the factors can be used for
an unsupervised analysis of samples or can be associated to biological
or technical covariates of the original samples. Additionally, the
reconstruction errors per view and factor can be used to prioritize cell
types associated with covariates of interest.</p>
</div>
<div class="section level2">
<h2 id="data">Data<a class="anchor" aria-label="anchor" href="#data"></a>
</h2>
<p>Here we show how to use MOFA for a multicellular factor analysis by
applying it to a cross-condition atlas.</p>
<p>As an example, we will use a toy dataset containing the pseudobulk
gene expression information of 22 samples across 3 cell types</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">inputs_dir</span> <span class="op">&lt;-</span> <span class="fu">base</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"MOFAcellulaR"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">inputs_dir</span>, <span class="st">"testpbcounts.rda"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">inputs_dir</span>, <span class="st">"testcoldata.rda"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">testcoldata</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">donor_id</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">donor_id</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 2</span></span></span>
<span><span class="co">#&gt;   donor_id      `n()`</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> GT/IZ_P13         3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> GT/IZ_P15         3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> GT/IZ_P9          3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> GT/IZ_P9_rep2     3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> IZ/BZ_P2          3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> IZ_P10            3</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">testcoldata</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">cell_type</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">cell_type</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 3 × 2</span></span></span>
<span><span class="co">#&gt;   cell_type `n()`</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> CM           22</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Endo         22</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> Fib          22</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="processing-pseudobulk-expression-profiles">1. Processing pseudobulk expression profiles<a class="anchor" aria-label="anchor" href="#processing-pseudobulk-expression-profiles"></a>
</h2>
<p>We will assume that regardless of the preferred way of storing your
pseudobulk data, a count matrix (genes in rows, samples in columns) will
be accompanied by the annotations of the columns that contain the
information of the cell type and sample of origin of the pseudobulk
expression vector.</p>
<p>If you are starting from a single cell data set, we recommend you to
use functions like <code><a href="https://rdrr.io/pkg/scuttle/man/summarizeAssayByGroup.html" class="external-link">scuttle::summarizeAssayByGroup()</a></code> to
generate objects as the ones used in this vignette.</p>
<p>In the example data <code>testpbcounts</code> contains a expression
matrix.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">testpbcounts</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;            CM_control_P1 Endo_control_P1 Fib_control_P1 CM_control_P17</span></span>
<span><span class="co">#&gt; AL627309.1             0               0              0              0</span></span>
<span><span class="co">#&gt; AL627309.5             0               0              0              0</span></span>
<span><span class="co">#&gt; AL627309.4             0               0              0              0</span></span>
<span><span class="co">#&gt; AL669831.2             0               0              0              0</span></span>
<span><span class="co">#&gt; LINC01409              1               1              1              1</span></span>
<span><span class="co">#&gt;            Endo_control_P17</span></span>
<span><span class="co">#&gt; AL627309.1                0</span></span>
<span><span class="co">#&gt; AL627309.5                0</span></span>
<span><span class="co">#&gt; AL627309.4                0</span></span>
<span><span class="co">#&gt; AL669831.2                0</span></span>
<span><span class="co">#&gt; LINC01409                 0</span></span></code></pre></div>
<p>And <code>testcoldata</code> contains the information of each sample
(column) of <code>testpbcounts</code> in a named
<code>data.frame</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">testcoldata</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;                     donor_id cell_type cell_counts</span></span>
<span><span class="co">#&gt; CM_control_P1     control_P1        CM          10</span></span>
<span><span class="co">#&gt; Endo_control_P1   control_P1      Endo          10</span></span>
<span><span class="co">#&gt; Fib_control_P1    control_P1       Fib          10</span></span>
<span><span class="co">#&gt; CM_control_P17   control_P17        CM          10</span></span>
<span><span class="co">#&gt; Endo_control_P17 control_P17      Endo          10</span></span>
<span><span class="co">#&gt; Fib_control_P17  control_P17       Fib          10</span></span></code></pre></div>
<p>The necessary components of <code>testcoldata</code> are the
<code>donor_id</code> column that refers to the sample of interest (eg.
patient), <code>cell_type</code> that will define the views in our
multicellular factor analysis, and <code>cell_counts</code> that will
allow to perform quality control filtering.</p>
<p><code>MOFAcellulaR</code> provides a series of useful tools to go
from these two data objects to a MOFA ready dataframe that we be used to
performed group factor analysis.</p>
<p>First, we create an initial <code>SummarizedExperiment</code> object
that will allow to make all the processing.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pb_obj</span> <span class="op">&lt;-</span> <span class="fu">MOFAcellulaR</span><span class="fu">::</span><span class="fu"><a href="../reference/create_init_exp.html">create_init_exp</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">testpbcounts</span>,  coldata <span class="op">=</span> <span class="va">testcoldata</span><span class="op">)</span></span></code></pre></div>
<p>Then, we will create a list of <code>SummarizedExperiment</code> of
samples and cell-types of interest.</p>
<p>An initial quality control step is to filter out pseudobulk samples
coming from a low number of cells, under the assumption that the gene
count estimates coming from a small population of cells may be
unreliable. The actual number of cells needed for a pseudobulk profile
is arbitrary and it is an empirical decision of the analyzer. If no
sample filtering is required, it is possible also to define that in the
helping functions.</p>
<p>In this example we only work with cardiomyocytes (CM) and fibroblasts
(Fib)</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ct_list</span> <span class="op">&lt;-</span> <span class="fu">MOFAcellulaR</span><span class="fu">::</span><span class="fu"><a href="../reference/filt_profiles.html">filt_profiles</a></span><span class="op">(</span>pb_dat <span class="op">=</span> <span class="va">pb_obj</span>,</span>
<span>                          cts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Fib"</span>,<span class="st">"CM"</span><span class="op">)</span>,</span>
<span>                          ncells <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                          counts_col <span class="op">=</span> <span class="st">"cell_counts"</span>, <span class="co"># This refers to the column name in testcoldata where the number of cells per profile was stored</span></span>
<span>                          ct_col <span class="op">=</span> <span class="st">"cell_type"</span><span class="op">)</span> <span class="co"># This refers to the column name in testcoldata where the cell-type label was stored</span></span></code></pre></div>
<p>The next step, requires to identify lowly expressed genes and highly
variable genes per cell-type independently. We reuse the same criteria
as <code>edgeR</code> to identify lowly expressed genes. Similarly as
the number of cells, this parameters should be decided by the
analyst.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ct_list</span> <span class="op">&lt;-</span> <span class="fu">MOFAcellulaR</span><span class="fu">::</span><span class="fu"><a href="../reference/filt_gex_byexpr.html">filt_gex_byexpr</a></span><span class="op">(</span>pb_dat_list <span class="op">=</span> <span class="va">ct_list</span>,</span>
<span>                            min.count <span class="op">=</span> <span class="fl">5</span>, <span class="co"># Modify!!</span></span>
<span>                            min.prop <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span> <span class="co"># Modify!!</span></span>
<span><span class="co">#&gt; Warning in filterByExpr.DGEList(y, design = design, group = group, lib.size =</span></span>
<span><span class="co">#&gt; lib.size, : All samples appear to belong to the same group.</span></span>
<span></span>
<span><span class="co">#&gt; Warning in filterByExpr.DGEList(y, design = design, group = group, lib.size =</span></span>
<span><span class="co">#&gt; lib.size, : All samples appear to belong to the same group.</span></span></code></pre></div>
<p>Since we override the group parameter, all samples are assumed to be
of the same group.</p>
<p>Normalization of pseudobulk expression profiles using Trimmed Mean of
the M-values (TMM) from <code><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html" class="external-link">edgeR::calcNormFactors</a></code> is then
performed</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ct_list</span> <span class="op">&lt;-</span> <span class="fu">MOFAcellulaR</span><span class="fu">::</span><span class="fu"><a href="../reference/tmm_trns.html">tmm_trns</a></span><span class="op">(</span>pb_dat_list <span class="op">=</span> <span class="va">ct_list</span>,</span>
<span>                                  scale_factor <span class="op">=</span> <span class="fl">1000000</span><span class="op">)</span></span></code></pre></div>
<p>Identification of highly variable genes per cell type is performed
with <code><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">scran::getTopHVGs()</a></code>, however you can also provide your
own list of highly variable genes if preferred</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ct_list</span> <span class="op">&lt;-</span> <span class="fu">MOFAcellulaR</span><span class="fu">::</span><span class="fu"><a href="../reference/filt_gex_byhvg.html">filt_gex_byhvg</a></span><span class="op">(</span>pb_dat_list <span class="op">=</span> <span class="va">ct_list</span>,</span>
<span>                                        prior_hvg <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                                        var.threshold <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p>Finally, it is possible to veto genes to be part of the model for
specific cell-types. The way we deal with this is to create a dictionary
of exclusive genes for a given cell-type. These could be for example
marker genes.</p>
<p>In this vignette, we will explicitly make TTN a gene exclusive for
cardiomyocytes, and POSTN exclusive to fibroblasts based on prior
knowledge, avoiding these to be background genes for other cell
types.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">prior_hvg_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"CM"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TTN"</span><span class="op">)</span>,</span>
<span>                        <span class="st">"Fib"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"POSTN"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ct_list</span> <span class="op">&lt;-</span> <span class="fu">MOFAcellulaR</span><span class="fu">::</span><span class="fu"><a href="../reference/filt_gex_bybckgrnd.html">filt_gex_bybckgrnd</a></span><span class="op">(</span>pb_dat_list <span class="op">=</span> <span class="va">ct_list</span>,</span>
<span>                                            prior_mrks <span class="op">=</span> <span class="va">prior_hvg_test</span><span class="op">)</span></span></code></pre></div>
<p>To convert the cell-type list into a MOFA ready object we just run
the following line</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">multiview_dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pb_dat2MOFA.html">pb_dat2MOFA</a></span><span class="op">(</span>pb_dat_list <span class="op">=</span> <span class="va">ct_list</span><span class="op">)</span></span></code></pre></div>
<p>All the previous steps can be concatenated using <code>%&gt;%</code>
for your convenience</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">multiview_dat</span> <span class="op">&lt;-</span> <span class="fu">MOFAcellulaR</span><span class="fu">::</span><span class="fu"><a href="../reference/create_init_exp.html">create_init_exp</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">testpbcounts</span>,  </span>
<span>                                               coldata <span class="op">=</span> <span class="va">testcoldata</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">MOFAcellulaR</span><span class="fu">::</span><span class="fu"><a href="../reference/filt_profiles.html">filt_profiles</a></span><span class="op">(</span>pb_dat <span class="op">=</span> <span class="va">.</span>,</span>
<span>                          cts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Fib"</span>,<span class="st">"CM"</span>, <span class="st">"Endo"</span><span class="op">)</span>,</span>
<span>                          ncells <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                          counts_col <span class="op">=</span> <span class="st">"cell_counts"</span>, <span class="co"># This refers to the column name in testcoldata where the number of cells per profile was stored</span></span>
<span>                          ct_col <span class="op">=</span> <span class="st">"cell_type"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">MOFAcellulaR</span><span class="fu">::</span><span class="fu"><a href="../reference/filt_gex_byexpr.html">filt_gex_byexpr</a></span><span class="op">(</span>pb_dat_list <span class="op">=</span> <span class="va">.</span>,</span>
<span>                                min.count <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                                min.prop <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">MOFAcellulaR</span><span class="fu">::</span><span class="fu"><a href="../reference/tmm_trns.html">tmm_trns</a></span><span class="op">(</span>pb_dat_list <span class="op">=</span> <span class="va">.</span>,</span>
<span>                         scale_factor <span class="op">=</span> <span class="fl">1000000</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">MOFAcellulaR</span><span class="fu">::</span><span class="fu"><a href="../reference/filt_gex_byhvg.html">filt_gex_byhvg</a></span><span class="op">(</span>pb_dat_list <span class="op">=</span> <span class="va">.</span>,</span>
<span>                               prior_hvg <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                               var.threshold <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">MOFAcellulaR</span><span class="fu">::</span><span class="fu"><a href="../reference/filt_gex_bybckgrnd.html">filt_gex_bybckgrnd</a></span><span class="op">(</span>pb_dat_list <span class="op">=</span> <span class="va">.</span>,</span>
<span>                                            prior_mrks <span class="op">=</span> <span class="va">prior_hvg_test</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">MOFAcellulaR</span><span class="fu">::</span><span class="fu"><a href="../reference/pb_dat2MOFA.html">pb_dat2MOFA</a></span><span class="op">(</span>pb_dat_list <span class="op">=</span> <span class="va">.</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="fitting-a-mofa-model">2. Fitting a MOFA model<a class="anchor" aria-label="anchor" href="#fitting-a-mofa-model"></a>
</h2>
<p>Once the single cell data is transformed into a multi-view
representation, now we can use MOFA to run a multicellular factor
analysis.</p>
<p>We will try to identify 6 factors that explain the variability
between patients captured by the seven different cell-types.</p>
<p>MOFA self-regularizes and will indicate a potential optimal number of
factors useful to describe the variability of your data, we advise to
follow the indications of <a href="https://biofam.github.io/MOFA2/tutorials.html" class="external-link">MOFA</a></p>
<p>Every factor captures coordination of gene expression across cell
types and will be called multicellular gene factors for the rest of the
vignette.</p>
<p>It is important to clarify what these factors capture:</p>
<ol style="list-style-type: lower-alpha">
<li>Coordinated expression of identical genes (generalistic response)
across cell-types</li>
<li>Coordinated expression of different genes (cell-type specific
response) across cell-types</li>
</ol>
<p>Fitting the model should take seconds.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MOFAobject</span> <span class="op">&lt;-</span> <span class="fu">MOFA2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MOFA2/man/create_mofa.html" class="external-link">create_mofa</a></span><span class="op">(</span><span class="va">multiview_dat</span><span class="op">)</span></span>
<span></span>
<span><span class="va">data_opts</span> <span class="op">&lt;-</span> <span class="fu">MOFA2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MOFA2/man/get_default_data_options.html" class="external-link">get_default_data_options</a></span><span class="op">(</span><span class="va">MOFAobject</span><span class="op">)</span></span>
<span><span class="va">train_opts</span> <span class="op">&lt;-</span> <span class="fu">MOFA2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MOFA2/man/get_default_training_options.html" class="external-link">get_default_training_options</a></span><span class="op">(</span><span class="va">MOFAobject</span><span class="op">)</span></span>
<span><span class="va">model_opts</span> <span class="op">&lt;-</span> <span class="fu">MOFA2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MOFA2/man/get_default_model_options.html" class="external-link">get_default_model_options</a></span><span class="op">(</span><span class="va">MOFAobject</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># This avoids the regularization of multicellular programs per cell type.</span></span>
<span><span class="co"># This avoids less sparse gene weights</span></span>
<span><span class="va">model_opts</span><span class="op">$</span><span class="va">spikeslab_weights</span> <span class="op">&lt;-</span> <span class="cn">FALSE</span> </span>
<span></span>
<span><span class="co"># Define the number of factors needed</span></span>
<span><span class="va">model_opts</span><span class="op">$</span><span class="va">num_factors</span> <span class="op">&lt;-</span> <span class="fl">5</span></span>
<span></span>
<span><span class="co"># Prepare MOFA model:</span></span>
<span><span class="va">MOFAobject</span> <span class="op">&lt;-</span> <span class="fu">MOFA2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MOFA2/man/prepare_mofa.html" class="external-link">prepare_mofa</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">MOFAobject</span>,</span>
<span>                           data_options <span class="op">=</span> <span class="va">data_opts</span>,</span>
<span>                           model_options <span class="op">=</span> <span class="va">model_opts</span>,</span>
<span>                           training_options <span class="op">=</span> <span class="va">train_opts</span><span class="op">)</span></span>
<span>  </span>
<span><span class="va">outfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"./vignettemodel.hdf5"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">model</span> <span class="op">&lt;-</span> <span class="fu">MOFA2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MOFA2/man/run_mofa.html" class="external-link">run_mofa</a></span><span class="op">(</span><span class="va">MOFAobject</span>, <span class="va">outfile</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; Warning in .quality_control(object, verbose = verbose): Factor(s) 1 are strongly correlated with the total number of expressed features for at least one of your omics. Such factors appear when there are differences in the total 'levels' between your samples, *sometimes* because of poor normalisation in the preprocessing steps.</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="exploring-the-mofa-model">3. Exploring the MOFA model<a class="anchor" aria-label="anchor" href="#exploring-the-mofa-model"></a>
</h2>
<div class="section level3">
<h3 id="exporting-model-outputs">Exporting model outputs<a class="anchor" aria-label="anchor" href="#exporting-model-outputs"></a>
</h3>
<p>For convenience, we provide functions to explore the results of your
model complementary to the ones already provided by <a href="https://biofam.github.io/MOFA2/tutorials.html" class="external-link">MOFA</a>
documentation.</p>
<p>These functions are based on the idea that users will have extra
information regarding the samples they analyzed. We provide supplemental
annotations of the toy object.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">inputs_dir</span>, <span class="st">"testmetadata.rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span></span>
<span><span class="co">#&gt;         sample patient_group batch</span></span>
<span><span class="co">#&gt; 1   control_P1      myogenic     A</span></span>
<span><span class="co">#&gt; 31    RZ/FZ_P5      myogenic     A</span></span>
<span><span class="co">#&gt; 61    RZ/BZ_P3      myogenic     A</span></span>
<span><span class="co">#&gt; 91       IZ_P3      ischemic     A</span></span>
<span><span class="co">#&gt; 121   RZ/BZ_P2      myogenic     A</span></span>
<span><span class="co">#&gt; 151      RZ_P3      myogenic     A</span></span></code></pre></div>
<p>Our sample meta data can also contain continous measurements if
needed, for example a clinical variable <code>fake_var</code></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">145</span><span class="op">)</span></span>
<span><span class="va">metadata</span><span class="op">$</span><span class="va">fake_var</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>You can obtain the factor scores of each of your samples by calling
the next function</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">all_factors</span> <span class="op">&lt;-</span> <span class="fu">MOFAcellulaR</span><span class="fu">::</span><span class="fu"><a href="../reference/get_tidy_factors.html">get_tidy_factors</a></span><span class="op">(</span>model <span class="op">=</span> <span class="va">model</span>,</span>
<span>                                 metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>                                 factor <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>                                 sample_id_column <span class="op">=</span> <span class="st">"sample"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">all_factors</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 6</span></span></span>
<span><span class="co">#&gt;   sample      patient_group batch fake_var Factor   value</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> control_P1  myogenic      A        0.687 Factor1 -<span style="color: #BB0000;">1.17</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> control_P1  myogenic      A        0.687 Factor2 -<span style="color: #BB0000;">0.519</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> control_P1  myogenic      A        0.687 Factor3 -<span style="color: #BB0000;">0.331</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> control_P1  myogenic      A        0.687 Factor4 -<span style="color: #BB0000;">0.612</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> control_P1  myogenic      A        0.687 Factor5  0.465</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> control_P17 myogenic      B       -<span style="color: #BB0000;">1.29</span>  Factor1 -<span style="color: #BB0000;">1.33</span></span></span></code></pre></div>
<p>You can specify also which factor you are interested and perform any
type of statistical analysis of interest</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Factor3</span> <span class="op">&lt;-</span> <span class="fu">MOFAcellulaR</span><span class="fu">::</span><span class="fu"><a href="../reference/get_tidy_factors.html">get_tidy_factors</a></span><span class="op">(</span>model <span class="op">=</span> <span class="va">model</span>,</span>
<span>                             metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>                             factor <span class="op">=</span> <span class="st">"Factor3"</span>,</span>
<span>                             sample_id_column <span class="op">=</span> <span class="st">"sample"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Factor3</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 6</span></span></span>
<span><span class="co">#&gt;   sample      patient_group batch fake_var Factor   value</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> control_P1  myogenic      A        0.687 Factor3 -<span style="color: #BB0000;">0.331</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> control_P17 myogenic      B       -<span style="color: #BB0000;">1.29</span>  Factor3 -<span style="color: #BB0000;">0.367</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> control_P7  myogenic      B        1.17  Factor3 -<span style="color: #BB0000;">0.151</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> control_P8  myogenic      B        0.793 Factor3 -<span style="color: #BB0000;">0.206</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> GT/IZ_P13   ischemic      B       -<span style="color: #BB0000;">0.354</span> Factor3  0.972</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> GT/IZ_P15   ischemic      B       -<span style="color: #BB0000;">1.02</span>  Factor3  0.363</span></span></code></pre></div>
<p>Each factor is composed by a linear combination of genes per
cell-type, and it is possible to extract the weights for each factor
with a practical function</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_weights</span> <span class="op">&lt;-</span> <span class="fu">MOFAcellulaR</span><span class="fu">::</span><span class="fu"><a href="../reference/get_geneweights.html">get_geneweights</a></span><span class="op">(</span>model <span class="op">=</span> <span class="va">model</span>, factor <span class="op">=</span> <span class="st">"Factor1"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gene_weights</span><span class="op">)</span></span>
<span><span class="co">#&gt;      feature       value ctype</span></span>
<span><span class="co">#&gt; 1     TTTY14 -0.07195983    CM</span></span>
<span><span class="co">#&gt; 2    TTN-AS1 -0.69494200    CM</span></span>
<span><span class="co">#&gt; 3 SLC8A1-AS1 -0.31866250    CM</span></span>
<span><span class="co">#&gt; 4     LRRTM3 -0.63194725    CM</span></span>
<span><span class="co">#&gt; 5   MLIP-AS1 -0.54372348    CM</span></span>
<span><span class="co">#&gt; 6  LINC02208 -0.82502754    CM</span></span></code></pre></div>
<p>These gene loadings are essential if you want to map cell-type
specific processes to bulk and spatial transcriptomics, since they can
be treated as gene sets. If you are interested in this, we refer to <a href="https://saezlab.github.io/decoupleR/#:~:text=decoupleR%20is%20a%20Bioconductor%20package,and%20weight%20of%20network%20interactions." class="external-link">decoupleR</a>
and <a href="https://decoupler-py.readthedocs.io/en/latest/" class="external-link">decoupler-py</a>
that explain in detail how to perform enrichment analysis with this type
of weighted gene sets.</p>
<p>Alternatively, the gene loadings can be reduced to functional or
cellular processes by enriching gene sets provided by literature in each
cell-type specific signature. Treat your gene loading matrix as scaled
transcriptomics and perform your enrichment test of preference, see
decoupleR’s documentation for this.</p>
</div>
<div class="section level3">
<h3 id="visualizing-sample-variability">Visualizing sample variability<a class="anchor" aria-label="anchor" href="#visualizing-sample-variability"></a>
</h3>
<p>As an initial exploratory analysis, one may want to visualize samples
in a 2D space, here we provide a plotting function that allows to
perform this using UMAPs or multidimensional-scaling plots.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">UMAP_embedding</span> <span class="op">&lt;-</span> <span class="fu">MOFAcellulaR</span><span class="fu">::</span><span class="fu"><a href="../reference/plot_sample_2D.html">plot_sample_2D</a></span><span class="op">(</span>model <span class="op">=</span> <span class="va">model</span>,</span>
<span>                                               method <span class="op">=</span> <span class="st">"UMAP"</span>,</span>
<span>                                               metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>                                               sample_id_column <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>                                               color_by <span class="op">=</span> <span class="st">"patient_group"</span><span class="op">)</span></span></code></pre></div>
<p><img src="get-started_files/figure-html/unnamed-chunk-21-1.png" width="384"></p>
</div>
<div class="section level3">
<h3 id="performing-statistical-analyses">Performing statistical analyses<a class="anchor" aria-label="anchor" href="#performing-statistical-analyses"></a>
</h3>
<p>To facilitate the exploration of the model, we provide a wrapper that
performs association tests between factor scores and covariates of the
samples, the covariates can be continuous or categorical. In the case of
continuous variables a linear model is fitted. For categorical
variables, analysis of variance (ANOVA) is performed.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">categorical_assoc</span> <span class="op">&lt;-</span> <span class="fu">MOFAcellulaR</span><span class="fu">::</span><span class="fu"><a href="../reference/get_associations.html">get_associations</a></span><span class="op">(</span>model <span class="op">=</span> <span class="va">model</span>,</span>
<span>                                      metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>                                       sample_id_column <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>                                       test_variable <span class="op">=</span> <span class="st">"patient_group"</span>,</span>
<span>                                       test_type <span class="op">=</span> <span class="st">"categorical"</span>,</span>
<span>                                       group <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">categorical_assoc</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 5 × 4</span></span></span>
<span><span class="co">#&gt;   Factor  term          p.value adj_pvalue</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Factor1 patient_group 0.009<span style="text-decoration: underline;">60</span>     0.048<span style="text-decoration: underline;">0</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Factor2 patient_group 0.072<span style="text-decoration: underline;">0</span>      0.216 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> Factor3 patient_group 0.315       0.547 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> Factor4 patient_group 0.037<span style="text-decoration: underline;">9</span>      0.152 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> Factor5 patient_group 0.274       0.547</span></span></code></pre></div>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">continuous_assoc</span> <span class="op">&lt;-</span> <span class="fu">MOFAcellulaR</span><span class="fu">::</span><span class="fu"><a href="../reference/get_associations.html">get_associations</a></span><span class="op">(</span>model <span class="op">=</span> <span class="va">model</span>,</span>
<span>                                       metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>                                       sample_id_column <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>                                       test_variable <span class="op">=</span> <span class="st">"fake_var"</span>,</span>
<span>                                       test_type <span class="op">=</span> <span class="st">"continuous"</span>,</span>
<span>                                       group <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">continuous_assoc</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 5 × 4</span></span></span>
<span><span class="co">#&gt;   Factor  term     p.value adj_pvalue</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Factor1 fake_var  0.018<span style="text-decoration: underline;">9</span>     0.094<span style="text-decoration: underline;">5</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Factor2 fake_var  0.455      1     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> Factor3 fake_var  0.254      1     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> Factor4 fake_var  0.561      1     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> Factor5 fake_var  0.825      1</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualizing-the-complete-model">Visualizing the complete model<a class="anchor" aria-label="anchor" href="#visualizing-the-complete-model"></a>
</h3>
<p>With MOFA you are able to interpret your model in distinct ways:</p>
<ol style="list-style-type: decimal">
<li><p>First it reduces the variability of samples across cell-types
inferring a latent space. See
<code><a href="../reference/get_tidy_factors.html">MOFAcellulaR::get_tidy_factors</a></code></p></li>
<li><p>The latent space captures certain percentage of the variability
of the original data, in this case the variability of samples within a
single cell-type. A full exploration of the model can be done
using:</p></li>
</ol>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">model</span><span class="op">@</span><span class="va">cache</span><span class="op">$</span><span class="va">variance_explained</span><span class="op">$</span><span class="va">r2_total</span></span>
<span><span class="co">#&gt; $single_group</span></span>
<span><span class="co">#&gt;       CM     Endo      Fib </span></span>
<span><span class="co">#&gt; 76.75793 67.30179 61.72416</span></span></code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li>Each latent variable contributes in explaining the variability of
the original data and that can also be used to prioritize signals.</li>
</ol>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">model</span><span class="op">@</span><span class="va">cache</span><span class="op">$</span><span class="va">variance_explained</span><span class="op">$</span><span class="va">r2_per_factor</span><span class="op">$</span><span class="va">single_group</span><span class="op">[</span>,,drop <span class="op">=</span> <span class="cn">F</span><span class="op">]</span></span>
<span><span class="co">#&gt;                CM       Endo       Fib</span></span>
<span><span class="co">#&gt; Factor1 42.440678 36.1886863 27.636917</span></span>
<span><span class="co">#&gt; Factor2 16.687453 20.8902217 20.198155</span></span>
<span><span class="co">#&gt; Factor3  9.965168  8.5866332  9.443090</span></span>
<span><span class="co">#&gt; Factor4  5.140941  0.8256951  2.962209</span></span>
<span><span class="co">#&gt; Factor5  2.498294  1.1868970  1.576150</span></span></code></pre></div>
<p>For example, in this model, we previously identified Factor1 to be
associated with our patient grouping, and based on the explaned
variance, we can say that while it represents a multicellular program of
the three cell-types we analyzed, Factor1 mainly captures the
variability of samples within CMs.</p>
<ol start="4" style="list-style-type: decimal">
<li>Each latent variable represents a multicellular program that can be
explored in detail with <code><a href="../reference/get_geneweights.html">MOFAcellulaR::get_geneweights</a></code>
</li>
</ol>
<p>To visualize the distinct components of the model, we provide a
heatmap plotting function that collects the distinct levels of results
of the model</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">assoc_list</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"categorical"</span> <span class="op">=</span> <span class="va">categorical_assoc</span>, <span class="st">"continuous"</span> <span class="op">=</span> <span class="va">continuous_assoc</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_MOFA_hmap.html">plot_MOFA_hmap</a></span><span class="op">(</span>model <span class="op">=</span> <span class="va">model</span>,</span>
<span>                group <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>                sample_id_column <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>                sample_anns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"patient_group"</span>, <span class="st">"batch"</span>, <span class="st">"fake_var"</span><span class="op">)</span>,</span>
<span>                assoc_list <span class="op">=</span> <span class="va">assoc_list</span><span class="op">)</span></span></code></pre></div>
<p><img src="get-started_files/figure-html/unnamed-chunk-26-1.png" width="432"></p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.2.3 (2023-03-15)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] grid      stats4    stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">#&gt; [8] methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] dplyr_1.1.1                 MOFAcellulaR_0.0.0.9000    </span></span>
<span><span class="co">#&gt;  [3] ggplot2_3.4.2               uwot_0.1.14                </span></span>
<span><span class="co">#&gt;  [5] Matrix_1.5-3                circlize_0.4.15            </span></span>
<span><span class="co">#&gt;  [7] ComplexHeatmap_2.14.0       MOFA2_1.8.0                </span></span>
<span><span class="co">#&gt;  [9] SummarizedExperiment_1.28.0 Biobase_2.58.0             </span></span>
<span><span class="co">#&gt; [11] GenomicRanges_1.50.2        GenomeInfoDb_1.34.9        </span></span>
<span><span class="co">#&gt; [13] IRanges_2.32.0              S4Vectors_0.36.2           </span></span>
<span><span class="co">#&gt; [15] BiocGenerics_0.44.0         MatrixGenerics_1.10.0      </span></span>
<span><span class="co">#&gt; [17] matrixStats_0.63.0         </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] Rtsne_0.16                  colorspace_2.1-0           </span></span>
<span><span class="co">#&gt;   [3] rjson_0.2.21                rprojroot_2.0.3            </span></span>
<span><span class="co">#&gt;   [5] scuttle_1.8.4               bluster_1.8.0              </span></span>
<span><span class="co">#&gt;   [7] XVector_0.38.0              BiocNeighbors_1.16.0       </span></span>
<span><span class="co">#&gt;   [9] GlobalOptions_0.1.2         fs_1.6.1                   </span></span>
<span><span class="co">#&gt;  [11] clue_0.3-64                 farver_2.1.1               </span></span>
<span><span class="co">#&gt;  [13] ggrepel_0.9.3               fansi_1.0.4                </span></span>
<span><span class="co">#&gt;  [15] codetools_0.2-19            sparseMatrixStats_1.10.0   </span></span>
<span><span class="co">#&gt;  [17] doParallel_1.0.17           cachem_1.0.7               </span></span>
<span><span class="co">#&gt;  [19] knitr_1.42                  jsonlite_1.8.4             </span></span>
<span><span class="co">#&gt;  [21] broom_1.0.4                 cluster_2.1.4              </span></span>
<span><span class="co">#&gt;  [23] png_0.1-8                   pheatmap_1.0.12            </span></span>
<span><span class="co">#&gt;  [25] HDF5Array_1.26.0            compiler_4.2.3             </span></span>
<span><span class="co">#&gt;  [27] dqrng_0.3.0                 basilisk_1.10.2            </span></span>
<span><span class="co">#&gt;  [29] backports_1.4.1             fastmap_1.1.1              </span></span>
<span><span class="co">#&gt;  [31] limma_3.54.2                cli_3.6.1                  </span></span>
<span><span class="co">#&gt;  [33] BiocSingular_1.14.0         htmltools_0.5.5            </span></span>
<span><span class="co">#&gt;  [35] tools_4.2.3                 igraph_1.4.2               </span></span>
<span><span class="co">#&gt;  [37] rsvd_1.0.5                  gtable_0.3.3               </span></span>
<span><span class="co">#&gt;  [39] glue_1.6.2                  GenomeInfoDbData_1.2.9     </span></span>
<span><span class="co">#&gt;  [41] reshape2_1.4.4              Rcpp_1.0.10                </span></span>
<span><span class="co">#&gt;  [43] jquerylib_0.1.4             pkgdown_2.0.7              </span></span>
<span><span class="co">#&gt;  [45] vctrs_0.6.1                 rhdf5filters_1.10.1        </span></span>
<span><span class="co">#&gt;  [47] iterators_1.0.14            DelayedMatrixStats_1.20.0  </span></span>
<span><span class="co">#&gt;  [49] xfun_0.38                   stringr_1.5.0              </span></span>
<span><span class="co">#&gt;  [51] beachmat_2.14.2             lifecycle_1.0.3            </span></span>
<span><span class="co">#&gt;  [53] irlba_2.3.5.1               statmod_1.5.0              </span></span>
<span><span class="co">#&gt;  [55] edgeR_3.40.2                zlibbioc_1.44.0            </span></span>
<span><span class="co">#&gt;  [57] scales_1.2.1                basilisk.utils_1.10.0      </span></span>
<span><span class="co">#&gt;  [59] ragg_1.2.5                  parallel_4.2.3             </span></span>
<span><span class="co">#&gt;  [61] rhdf5_2.42.1                RColorBrewer_1.1-3         </span></span>
<span><span class="co">#&gt;  [63] SingleCellExperiment_1.20.1 yaml_2.3.7                 </span></span>
<span><span class="co">#&gt;  [65] memoise_2.0.1               reticulate_1.28            </span></span>
<span><span class="co">#&gt;  [67] sass_0.4.5                  stringi_1.7.12             </span></span>
<span><span class="co">#&gt;  [69] highr_0.10                  desc_1.4.2                 </span></span>
<span><span class="co">#&gt;  [71] corrplot_0.92               foreach_1.5.2              </span></span>
<span><span class="co">#&gt;  [73] ScaledMatrix_1.6.0          scran_1.26.2               </span></span>
<span><span class="co">#&gt;  [75] filelock_1.0.2              BiocParallel_1.32.6        </span></span>
<span><span class="co">#&gt;  [77] shape_1.4.6                 rlang_1.1.0                </span></span>
<span><span class="co">#&gt;  [79] pkgconfig_2.0.3             systemfonts_1.0.4          </span></span>
<span><span class="co">#&gt;  [81] bitops_1.0-7                evaluate_0.20              </span></span>
<span><span class="co">#&gt;  [83] lattice_0.20-45             purrr_1.0.1                </span></span>
<span><span class="co">#&gt;  [85] Rhdf5lib_1.20.0             labeling_0.4.2             </span></span>
<span><span class="co">#&gt;  [87] cowplot_1.1.1               tidyselect_1.2.0           </span></span>
<span><span class="co">#&gt;  [89] plyr_1.8.8                  magrittr_2.0.3             </span></span>
<span><span class="co">#&gt;  [91] R6_2.5.1                    generics_0.1.3             </span></span>
<span><span class="co">#&gt;  [93] metapod_1.6.0               DelayedArray_0.24.0        </span></span>
<span><span class="co">#&gt;  [95] pillar_1.9.0                withr_2.5.0                </span></span>
<span><span class="co">#&gt;  [97] RCurl_1.98-1.12             tibble_3.2.1               </span></span>
<span><span class="co">#&gt;  [99] dir.expiry_1.6.0            crayon_1.5.2               </span></span>
<span><span class="co">#&gt; [101] utf8_1.2.3                  rmarkdown_2.21             </span></span>
<span><span class="co">#&gt; [103] GetoptLong_1.0.5            locfit_1.5-9.7             </span></span>
<span><span class="co">#&gt; [105] FNN_1.1.3.2                 forcats_1.0.0              </span></span>
<span><span class="co">#&gt; [107] digest_0.6.31               tidyr_1.3.0                </span></span>
<span><span class="co">#&gt; [109] textshaping_0.3.6           munsell_0.5.0              </span></span>
<span><span class="co">#&gt; [111] bslib_0.4.2</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Ricardo O. Ramirez Flores.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
